python/ray/data/_internal/plan.py

```diff
@@ -352,62 +355,33 @@ def schema(
         Returns:
             The schema of the output dataset.
         """
-        from ray.data._internal.logical.operators.all_to_all_operator import (
-            RandomizeBlocks,
-        )
-
         if self._schema is not None:
+            print("Using cached schema")
             return self._schema
 
-        if self._snapshot_blocks is None or (
-            self._snapshot_operator is not None
-            and self._snapshot_operator.output_dependencies
+        if (
+            self._snapshot_bundle is not None
+            and not self._snapshot_operator.output_dependencies
         ):
-            # There remain some operators yet to be executed.
-            # Even if a schema is already previously known, it may change,
-            # so we try executing to get the most updated schema.
-
-            # TODO(swang): There are several other operator types that could
-            # inherit the schema or we can compute the schema without having to
-            # execute any of the dataset: limit, filter, map_batches for
-            # add/drop columns, etc.
-            if fetch_if_missing:
-                if isinstance(self._logical_plan.dag, RandomizeBlocks):
-                    # TODO(ekl): this is a hack to optimize the case where we have a
-                    # trailing randomize block operator. That operator has no effect and
-                    # so we don't need to execute all blocks to get the schema.
-
-                    randomize_blocks_op = self._logical_plan.dag
-                    self._logical_plan._dag = randomize_blocks_op.input_dependencies[0]
-                    try:
-                        self.execute()
-                    finally:
-                        self._logical_plan = randomize_blocks_op
```

Comment:
I'm removing this optimization, because I don't think users call `randomize_blocks` often, and I don't think this performance improvement from this hack is worth the additional code complexity.